File: java/02-intermediate/06-exception-handling/ExceptionHandling.java
import java.io.*;

/**
 * LESSON: Exception Handling
 * Exceptions are errors that occur during runtime.
 * Handle them gracefully instead of crashing.
 *
 * Checked exceptions   - must be handled (IOException, SQLException)
 * Unchecked exceptions  - optional to handle (NullPointerException, ArithmeticException)
 */
public class ExceptionHandling {
    public static void main(String[] args) {
        // ===== TRY-CATCH =====
        System.out.println("--- Basic Try-Catch ---");
        try {
            int result = 10 / 0; // ArithmeticException
            System.out.println(result); // never reached
        } catch (ArithmeticException e) {
            System.out.println("Error: " + e.getMessage());
        }

        // ===== MULTIPLE CATCH BLOCKS =====
        System.out.println("\n--- Multiple Catch ---");
        try {
            int[] arr = {1, 2, 3};
            System.out.println(arr[10]); // ArrayIndexOutOfBoundsException
        } catch (ArrayIndexOutOfBoundsException e) {
            System.out.println("Array error: " + e.getMessage());
        } catch (Exception e) {
            System.out.println("General error: " + e.getMessage());
        }

        // ===== TRY-CATCH-FINALLY =====
        System.out.println("\n--- Finally Block ---");
        try {
            int x = Integer.parseInt("abc"); // NumberFormatException
        } catch (NumberFormatException e) {
            System.out.println("Parse error: " + e.getMessage());
        } finally {
            // ALWAYS executes, even if exception occurs or not
            System.out.println("Finally block always runs!");
        }

        // ===== THROW KEYWORD =====
        System.out.println("\n--- Throw ---");
        try {
            validateAge(15);
        } catch (IllegalArgumentException e) {
            System.out.println("Validation error: " + e.getMessage());
        }

        // ===== CUSTOM EXCEPTION =====
        System.out.println("\n--- Custom Exception ---");
        try {
            withdraw(100, 500);
        } catch (InsufficientFundsException e) {
            System.out.println("Bank error: " + e.getMessage());
            System.out.println("Short by: $" + e.getShortage());
        }

        // ===== TRY-WITH-RESOURCES =====
        System.out.println("\n--- Try-With-Resources ---");
        // Automatically closes resources (anything implementing AutoCloseable)
        try (BufferedWriter writer = new BufferedWriter(new FileWriter("test_output.txt"))) {
            writer.write("Hello from try-with-resources!");
            System.out.println("File written successfully");
        } catch (IOException e) {
            System.out.println("IO Error: " + e.getMessage());
        }
        // writer is automatically closed here, even if exception occurred
    }

    // 'throws' declares that this method might throw an exception
    static void validateAge(int age) {
        if (age < 18) {
            throw new IllegalArgumentException("Age must be 18+. Got: " + age);
        }
        System.out.println("Age is valid: " + age);
    }

    static void withdraw(double balance, double amount) throws InsufficientFundsException {
        if (amount > balance) {
            throw new InsufficientFundsException(amount - balance);
        }
        System.out.println("Withdrawal successful. Remaining: $" + (balance - amount));
    }
}

// Custom exception class
class InsufficientFundsException extends Exception {
    private double shortage;

    InsufficientFundsException(double shortage) {
        super("Insufficient funds in account");
        this.shortage = shortage;
    }

    double getShortage() { return shortage; }
}

File: cpp-practice/02-intermediate/06-exception-handling/exception_handling.cpp
/**
 * LESSON: Exception Handling
 * Handle runtime errors gracefully with try-catch-throw.
 */
#include <iostream>
#include <string>
#include <stdexcept> // standard exception classes
#include <vector>

// Custom exception class
class InsufficientFundsException : public std::runtime_error {
    double shortage;
public:
    InsufficientFundsException(double shortage)
        : std::runtime_error("Insufficient funds"), shortage(shortage) {}
    double getShortage() const { return shortage; }
};

class NegativeValueException : public std::invalid_argument {
public:
    NegativeValueException(const std::string& msg)
        : std::invalid_argument(msg) {}
};

double divide(double a, double b) {
    if (b == 0) throw std::runtime_error("Division by zero!");
    return a / b;
}

void withdraw(double balance, double amount) {
    if (amount < 0) throw NegativeValueException("Amount cannot be negative");
    if (amount > balance) throw InsufficientFundsException(amount - balance);
    std::cout << "Withdrawal successful. Remaining: $" << (balance - amount) << "\n";
}

int main() {
    // ===== BASIC TRY-CATCH =====
    std::cout << "--- Basic Try-Catch ---\n";
    try {
        double result = divide(10, 0);
        std::cout << result << "\n"; // never reached
    } catch (const std::runtime_error& e) {
        std::cout << "Error: " << e.what() << "\n";
    }

    // ===== MULTIPLE CATCH BLOCKS =====
    std::cout << "\n--- Multiple Catch ---\n";
    try {
        std::vector<int> v = {1, 2, 3};
        std::cout << v.at(10) << "\n"; // throws std::out_of_range
    } catch (const std::out_of_range& e) {
        std::cout << "Out of range: " << e.what() << "\n";
    } catch (const std::exception& e) {
        std::cout << "General error: " << e.what() << "\n";
    }

    // ===== CUSTOM EXCEPTIONS =====
    std::cout << "\n--- Custom Exceptions ---\n";
    try {
        withdraw(100, 500);
    } catch (const InsufficientFundsException& e) {
        std::cout << "Bank error: " << e.what() << "\n";
        std::cout << "Short by: $" << e.getShortage() << "\n";
    }

    try {
        withdraw(100, -50);
    } catch (const NegativeValueException& e) {
        std::cout << "Validation error: " << e.what() << "\n";
    }

    // ===== CATCH ALL =====
    std::cout << "\n--- Catch All ---\n";
    try {
        throw 42; // throwing a non-standard type
    } catch (...) { // catches anything
        std::cout << "Caught unknown exception\n";
    }

    // ===== NESTED TRY-CATCH =====
    std::cout << "\n--- Nested Try-Catch ---\n";
    try {
        try {
            throw std::runtime_error("Inner error");
        } catch (const std::runtime_error& e) {
            std::cout << "Inner catch: " << e.what() << "\n";
            throw; // re-throw to outer handler
        }
    } catch (const std::exception& e) {
        std::cout << "Outer catch: " << e.what() << "\n";
    }

    // ===== STANDARD EXCEPTION HIERARCHY =====
    std::cout << "\n--- Standard Exceptions ---\n";
    std::cout << "std::exception\n";
    std::cout << "  ├── std::runtime_error\n";
    std::cout << "  │   ├── std::overflow_error\n";
    std::cout << "  │   ├── std::underflow_error\n";
    std::cout << "  │   └── std::range_error\n";
    std::cout << "  ├── std::logic_error\n";
    std::cout << "  │   ├── std::invalid_argument\n";
    std::cout << "  │   ├── std::out_of_range\n";
    std::cout << "  │   └── std::domain_error\n";
    std::cout << "  └── std::bad_alloc (new fails)\n";

    // ===== NOEXCEPT =====
    std::cout << "\n--- noexcept ---\n";
    // Mark functions that won't throw (helps compiler optimize)
    auto safeAdd = [](int a, int b) noexcept { return a + b; };
    std::cout << "safeAdd: " << safeAdd(3, 4) << "\n";

    return 0;
}
